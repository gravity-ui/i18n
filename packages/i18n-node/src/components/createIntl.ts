import type {FallbackLocales, Messages} from '@gravity-ui/i18n-types';
import {
    createIntl as coreCreateIntl,
    createIntlCache,
    getFallbackLocalesList,
    getFormatters,
    getIntlFormatters,
    isFallbackOnEmptyString,
    memoize,
    onErrorFactory,
    validateLocale,
} from '@gravity-ui/i18n-core';

import type {
    IntlConfig,
    IntlShape,
    IntlShapeWithLocale,
    ResolvedIntlConfigWithLocale,
} from '../types';

import {factoryCreateMessagesFunction} from './factoryCreateMessagesFunction';

export function createIntl<AvailableLocale extends string>(
    config: IntlConfig<AvailableLocale>,
): IntlShape<AvailableLocale> {
    const intlShapeConfig: Omit<
        IntlShape<AvailableLocale>,
        'getLocaleInstance' | 'createMessages'
    > = {
        allowedLocales: config.allowedLocales,
        disableUseLocaleLangAsFallback: config.disableUseLocaleLangAsFallback ?? false,
        fallbackLocales: config.fallbackLocales ?? ({} as FallbackLocales<AvailableLocale>),
        defaultFallback: config.defaultFallback ?? 'empty-string',
    };

    const createMessages = <K extends string>(messages: Messages<AvailableLocale, K>) => ({
        messages,
    });

    const getLocaleInstance = memoize(
        (locale: AvailableLocale | string): IntlShapeWithLocale<AvailableLocale> => {
            const cache = createIntlCache();
            const coreIntl = coreCreateIntl(
                {
                    defaultLocale: locale,
                    fallbackOnEmptyString: isFallbackOnEmptyString(config),
                    locale,
                    ...config,
                    ...intlShapeConfig,
                    onError: onErrorFactory({onError: config.onError}),
                },
                cache,
            );

            const resolvedIntlConfig: ResolvedIntlConfigWithLocale<AvailableLocale> = {
                defaultRichTextElements: coreIntl.defaultRichTextElements,
                onError: coreIntl.onError,
                onWarn: coreIntl.onWarn,

                getLocale: () => locale,

                formats: {
                    number: coreIntl.formats.number,
                },
                defaultFormats: {
                    number: coreIntl.defaultFormats.number,
                },

                allowedLocales: config.allowedLocales,
                fallbackLocales: config.fallbackLocales ?? ({} as FallbackLocales<AvailableLocale>),
                defaultFallback: config.defaultFallback ?? 'empty-string',
                disableUseLocaleLangAsFallback: config.disableUseLocaleLangAsFallback ?? false,
                getCurrentFallbackLocales: function () {
                    return getFallbackLocalesList(locale, this);
                },
            };

            if (!validateLocale(locale, resolvedIntlConfig)) {
                // TODO It crashes page, maybe other behavior
                // TODO Add message
                throw new Error('');
            }

            const intlFormatters = getIntlFormatters(coreIntl);
            const formatters = getFormatters(coreIntl.formatters);

            const useMessages = factoryCreateMessagesFunction(resolvedIntlConfig, {
                formatMessage: coreIntl.formatMessage,
            });

            return {
                ...resolvedIntlConfig,
                ...intlFormatters,
                formatters,
                useMessages,
            };
        },
    );

    return {
        ...intlShapeConfig,
        getLocaleInstance,
        createMessages,
    };
}
