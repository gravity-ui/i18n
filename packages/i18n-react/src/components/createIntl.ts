import type React from 'react';

import {
    clearCacheOnPlace,
    getFallbackLocalesList,
    getFormatters,
    getIntlFormatters,
    isFallbackOnEmptyString,
    onErrorFactory,
    validateLocale,
} from '@gravity-ui/i18n-core';
import {createIntl as coreCreateIntl, createIntlCache} from 'react-intl';

import type {IntlConfig, IntlShape, ResolvedIntlConfig} from '../types';

import {factoryCreateMessagesFunction} from './factoryCreateMessagesFunction';

export function createIntl<AvailableLocale extends string>(
    config: IntlConfig<AvailableLocale>,
): IntlShape<AvailableLocale> {
    let currentLocale = config.locale;

    const cache = createIntlCache();
    const coreIntl = coreCreateIntl(
        {
            defaultLocale: currentLocale,
            fallbackOnEmptyString: isFallbackOnEmptyString(config),
            ...config,
            onError: onErrorFactory({onError: config.onError}),
        },
        cache,
    );

    const resolvedIntlConfig: ResolvedIntlConfig<AvailableLocale> = {
        defaultRichTextElements: coreIntl.defaultRichTextElements,
        onError: coreIntl.onError,
        onWarn: coreIntl.onWarn,

        getLocale: () => currentLocale,

        formats: {
            number: coreIntl.formats.number,
        },
        defaultFormats: {
            number: coreIntl.defaultFormats.number,
        },

        allowedLocales: config.allowedLocales,
        fallbackLocales:
            config.fallbackLocales ??
            ({} as ResolvedIntlConfig<AvailableLocale>['fallbackLocales']),
        defaultFallback: config.defaultFallback ?? 'empty-string',
        disableUseLocaleLangAsFallback: config.disableUseLocaleLangAsFallback ?? false,
        getCurrentFallbackLocales: function () {
            return getFallbackLocalesList(currentLocale, this);
        },

        textComponent: coreIntl.textComponent,
        wrapRichTextChunksInFragment: coreIntl.wrapRichTextChunksInFragment,
    };

    if (!validateLocale(currentLocale, resolvedIntlConfig)) {
        // TODO It crashes page, maybe other behavior
        // TODO Add message
        throw new Error('');
    }

    const createMessages = factoryCreateMessagesFunction(resolvedIntlConfig, {
        formatMessage: coreIntl.formatMessage,
    });

    const intlFormatters = getIntlFormatters<React.ReactNode>(coreIntl);
    const formatters = getFormatters(coreIntl.formatters);

    const setLocale = (locale: AvailableLocale | string) => {
        if (locale === currentLocale) {
            return;
        }

        if (!validateLocale(locale, resolvedIntlConfig)) {
            return;
        }

        clearCacheOnPlace(cache);
        currentLocale = locale;
        coreIntl.locale = locale;
        coreIntl.defaultLocale = locale;
    };

    return {
        ...resolvedIntlConfig,
        ...intlFormatters,
        formatters,
        createMessages,
        setLocale,
    };
}
