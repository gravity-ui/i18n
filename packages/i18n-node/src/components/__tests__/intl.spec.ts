import {createIntl} from '../createIntl';

describe('intl entity', () => {
    it('should create intl entity with valid config', () => {
        expect(
            createIntl({
                allowedLocales: ['en'],
            }),
        ).not.toBeUndefined();
    });

    it('should create messages without synchronization', () => {
        const create = 'Create';

        const intl = createIntl({
            allowedLocales: ['en'],
        });

        const {messages} = intl.createMessages({
            create: {
                en: create,
            },
        });

        const instance = intl.getLocaleInstance('en');
        const {t} = instance.useMessages(messages);

        expect(t('create')).toBe(create);
    });

    it('should create messages with synchronization', () => {
        const create = 'Create';

        const intl = createIntl({
            allowedLocales: ['en'],
        });

        const {messages} = intl.createMessages({
            create: {
                en: create,
                meta: {
                    id: 'create-id',
                },
            },
        });

        const instance = intl.getLocaleInstance('en');
        const {t} = instance.useMessages(messages);

        expect(t('create')).toBe(create);
    });
});

describe('multi-locale', () => {
    it('should allow change locales', () => {
        const createEN = 'Create';
        const createRU = 'Создать';

        const intl = createIntl({
            allowedLocales: ['en', 'ru'],
        });

        const {messages} = intl.createMessages({
            create: {
                en: createEN,
                ru: createRU,
            },
        });

        const instanceEN = intl.getLocaleInstance('en');
        const {t: tEN} = instanceEN.useMessages(messages);
        expect(instanceEN.getLocale()).toBe('en');
        expect(tEN('create')).toBe(createEN);

        const instanceRU = intl.getLocaleInstance('ru');
        const {t: tRU} = instanceRU.useMessages(messages);
        expect(instanceRU.getLocale()).toBe('ru');
        expect(tRU('create')).toBe(createRU);
    });

    it('should throw error if new locale is invalid', () => {
        let error;
        let errorFromOnError;

        const intl = createIntl({
            allowedLocales: ['en'],
            onError: (e) => {
                errorFromOnError = e;
            },
        });

        try {
            intl.getLocaleInstance('ru');
        } catch (e) {
            error = e;
        }

        expect(String(error)).toMatchSnapshot('set invalid locale error');
        expect(String(errorFromOnError)).toMatchSnapshot('set invalid locale error from onError');
    });

    it('should change fallback locales after change locale', () => {
        const intl = createIntl({
            allowedLocales: ['en', 'en-kz', 'ru', 'ru-kz'],
            fallbackLocales: {
                'en-kz': 'en',
                'ru-kz': 'ru',
            },
        });

        const instanceEN = intl.getLocaleInstance('en-kz');
        expect(instanceEN.getCurrentFallbackLocales()).toEqual(['en']);

        const instanceRU = intl.getLocaleInstance('ru-kz');
        expect(instanceRU.getCurrentFallbackLocales()).toEqual(['ru']);
    });
});
