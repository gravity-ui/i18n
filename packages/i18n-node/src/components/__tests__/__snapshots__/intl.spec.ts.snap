// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multi-locale should throw error if new locale is invalid: set invalid locale error 1`] = `"Error"`;

exports[`multi-locale should throw error if new locale is invalid: set invalid locale error from onError 1`] = `
"Error: [@formatjs/intl Error MISSING_DATA] Missing locale data for locale: "ru" in config.allowedLocales and in config.fallbackLocales.
"
`;
